csharp/src/SeedLang/IEngine.cs

```diff
@@ -0,0 +1,37 @@
+// Copyright 2021 The Aha001 Team.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+using SeedLang.Common;
+
+namespace SeedLang {
+  // The interface of the SeedLang engine.
+  public interface IEngine {
+    // A collection that contains all binary operators.
+    IEnumerable<string> BinaryOperators { get; }
+
+    // Runs source code with the specified module name, language, target parsing rule and runing
+    // type.
+    //
+    // There are three mothods to run the source code.
+    // 1) Dryrun: parses and validates the source code without runing.
+    // 2) Source: parses the source code into an AST tree, and runs it by traversing the AST tree.
+    // 3) Bytecode: parses and compiles the source code into bytecode, and runs it in a VM.
+    //
+    // The parsing and running diagnostic information will be collected into the diagnostic
+    // collection.
+    bool Run(string source, string module, Language language, ParseRule rule, RunType runType,
```

Comment:
We have decided via emails that the dependency chain will be:

Shell / Unity UI -> Engine -> parsers implemented by SeedBlock / SeedX -> SeedAst

Meanwhile, there will obviously be another dependency chain for the UI to operate on in-memory blocks:

Unity UI -> SeedBlock

Given the above dependency chains, a Unity UI can parse a SeedBlock code into an Engine instance to parse the code. Hence, the interface Run() here need to accept an in-memory SeedBlock module object and parse it to SeedBlock's parser.

It results in a new cycle dependency if we keep the things as they are.

Probably we need to extract the interface of the in-memory block (the BaseBlock interface) out to SeedLang.Common?

Your thoughts?

Reply:
The dependency should be as following:

```
                                 Unity UI
                          -----------------
                          |                            |
                          |                            |
                         V                          V
               SeedBlock. <------ Engine -------> X
                         |                                                     |
                         |                                                     |
                        --------------------------------
                                                |
                                               V
                                             Ast

```

Reply:
Ah... my bad... I mistakenly mixed something in my mind when I was reading your code. Your diagram is pretty clear. Forget my comments.